tokio/docs/runtime-stats.md

```diff
@@ -0,0 +1,137 @@
+
+### I/O driver
+
+Unlike, the executor stats, stats coming from the I/O driver will be streamed directly to the `Stats` struct via atomics. Each value will be incremented (via `AtomicU64::fetch_add`) for each event.
+
+List of stats provided from the I/O driver:
+- Amount of compact
+- Amount of "token dispatches" (aka ready events)
+- Amount of fd currently registered with `io::Driver`
+- Amount of read ready events
+- Amount of write ready events
+
+### Timer driver
+
+Similar to the I/O driver the timer driver will stream stats in a similar fashion. These will be emitted via `AtomicU64::fetch_add` and be emitted on each occurance of the events below.
+
+List of stats provided from the timer driver:
+- Amount of timer entries (timer wheel depth)
+- Amount of timers fired
+- Amount of timers cancelled
+
+### `tokio-metrics`
+
+The `tokio-metrics` crate will provide aggregated metrics based on the `Stats` struct. This will include histograms and other useful aggregated forms of the stats that could be emitted by various metrics implementations. This crate is designed to provide the ability to expose the aggregated stats in an unstable `0.1` method outside of the runtime and allow the ability to iterate on how they are aggregated without the need to follow `tokio`'s strict versioning scheme.
+
+The rest of this crate is out of scope for this RFC, since most of the implementation will happen ad-hoc outside of the main crate.
+
+## Implementation details
+
+### Batch streaming
+
+To avoid any extra overhead in the executor loop, each worker will batch metrics into a `Core` local struct. These values will be incremented or sampled during regular executor cycles when certain operations happen like a work steal attempt or a pop from one of the queues.
+
+The batches will be streamed via atomics to the stats struct directly. This will reduce any cross CPU work while the executor is running and amortize the cost of having to do cross CPU work. Batches will be sent before the executor attempts to park the thread. This will happen either when there is no work to be done or when the executor has hit the maintance tick. At this point before the thread will park, the executor will submit the batch. Generally, since parking is more expensive then submitting batches there should not be any added latency to the executor cycle in this process.
+
+This then allows the collector to poll the stats on any interval. Thus, allowing it to drive its own timer to understand an estimate of the duration that a certain amount of ticks too, or how many times the the executor entered the park state.
+
+### I/O and Timer implementations
+
+Since, the two drivers (I/O and timer) that `tokio` provides are singtons within the runtime there is no need to iterate through their stats like the executor stats. In-addition, it is possible to stream the metrics directly from the driver events rather than needing to batch them like the executor.
```

Comment:
When we are using a lot of atomics in this manner, we should be careful regarding false sharing of the atomics.

Reply:
Can you expand on what you mean by false sharing?

Reply:
If you have a bunch of atomic variables stored together (in the same cache line), with many threads writing to them concurrently, then this can impact performance quite a lot, even if the writes are affecting two different counters.

Reply:
In this case the issue is true sharing (contention on atomics), so padding stuff out won't solve it either.

Reply:
How is it a case of true sharing? My read is that there is a single driver and one thread polls it. It can store stats in an atomic and an arbitrary number of stats aggregators can load it.

False sharing only happens when there are concurrent mutations, which is not (afaik) the case here.